### DISCUSSION TITLE
Difference between `self._sliced_trajectory` and `self._trajectory` in analysis classes

### DISCUSSION THREAD

**Discussion by marinegor** [MDAnalysis/mdanalysis#4426] [Development]
Hi everyone,

I wonder what's the difference between `self._trajectory` (that gets initialized in `AnalysisBase.__init__`) and `self._sliced_trajectory` (initialized in `_setup_frames`), and which should be used in which case?

The way I see it, is that `self._trajectory` should basically never be used in analysis runs, since it doesn't hold the information about `start, stop, step` or `frames` arguments, passed to `run()` method. However, I sometimes see subclasses using directly `self._trajectory` (e.g. `diffusionmap.DistanceMatrix` or `pca.PCA`). What's the rationale for that?

Tagging @hmacdope and @IAlibay since it continues the discussion in PR #4423

**Comment by IAlibay** [comment]
Off the top of my head, the likely reasoning here is that `self._sliced_trajectory` is a new addition to `AnalysisBase` (well 2 years new), which means that a lot of the analysis classes haven't been switched over.

I'm going to tag @orbeckst here who might have a better recollection of all this.

**Comment by orbeckst** [comment]
I think you're both correct.

Did you look at the git-blame log to find out where we implemented the change? The corresponding PR might have additional information.

We should document internal usage — comments and docs.

**Comment by marinegor** [comment]
It was @rsexton2 in #3415 but I can't see the description of `_sliced_trajectory` (and reasons why other classes don't use it).

I now assume that all subclasses should not touch `self._trajectory` at all, right? Meaning, all these:

<img width="1032" alt="image" src="https://github.com/MDAnalysis/mdanalysis/assets/20686002/298aac60-b14f-4ed5-9255-c9f63e79a50b">

and these:

<img width="1003" alt="image" src="https://github.com/MDAnalysis/mdanalysis/assets/20686002/a984c687-5bf6-4ede-bf56-0665393339d9">


And this is probably the reason for the bug #4423 (and few others when analysis is run with explicit `frames`)?

**Reply by orbeckst** [reply]
I think I agree with your assessment.

Looking at #3415, it looks to me that other analysis tools _should_ be using `self._sliced_trajectory` instead of `self._trajectory[start:stop:step]` to be able to use `frames` or `start/stop/step`. This probably only affects analysis classes that do more than implementing `_single_frame()` and instead do their own thing.

**Reply by marinegor** [reply]
Ok, so I'll probably add a separate PR fixing it in all but PCA subclasses, after #4423 is merged.

**Reply by marinegor** [reply]
Separate PR added: #4433 (turned out only `diffusionmap` needs that)
